Сверстать страницы "Cегменты", "Сегмент" и "Добавление нового сегмента"

Beken-S opened the pull request:
**Добавил:**
- написал модели с моковыми данными
- сделали компонен таблицы для сегментов
- сделал компонент формы для добавления сегмента
- добавил лейаут для форм
- добавил пункт "Cегменты" в NavMenu
- сделал страницы "Cегменты", "Сегмент" и "Добавление нового сегмента"
- добавил хелпер sleep для имитации загрузки
- добавил хелпер isMessage для типизации сообщений ошибок React Final Form

**Исправил и изменил:**
- сделал строку поиска более похожей на строку поиска из макета
- исправил расположение строки поиска, заголовков и кнопок для добавления на страницах Слоев, Полей, Экранов
- переименовал часть файлов
- добавил локализацию в тему
- исправил отображение цвета у кнопки удаления в LightenChip
- переименовал пропс dataConverter в dataRender и добавил возможность пробрасывать экшен-функцию TableDataRow

**Страница "Сегменты"**

*загрузка*
![Снимок экрана от 2023-05-25 02-59-45](https://github.com/hhru-school/segment-admin-panel-front/assets/84242545/cd2448c4-9398-4015-a6d9-7b7996e31768)

![Снимок экрана от 2023-05-25 02-58-11](https://github.com/hhru-school/segment-admin-panel-front/assets/84242545/516a823f-8791-49aa-bc80-e4a806041a38)

**Страница "Сегмент"**

*загрузка*
![Снимок экрана от 2023-05-25 03-00-18](https://github.com/hhru-school/segment-admin-panel-front/assets/84242545/93452974-9bd4-4cf2-af00-26449d73190a)

![Снимок экрана от 2023-05-25 03-00-22](https://github.com/hhru-school/segment-admin-panel-front/assets/84242545/8e375184-9649-483a-a8f2-c12e7a4ced46)

**Форма добавления сегмента**

*загрузка*
![Снимок экрана от 2023-05-25 02-58-15](https://github.com/hhru-school/segment-admin-panel-front/assets/84242545/4e749a02-513b-4af6-af42-293bd666ba4e)

*заполнение*
![Снимок экрана от 2023-05-25 02-59-39](https://github.com/hhru-school/segment-admin-panel-front/assets/84242545/092d320e-9b93-4c15-ad23-5c8647f594d5)
![Снимок экрана от 2023-05-25 03-00-00](https://github.com/hhru-school/segment-admin-panel-front/assets/84242545/144b3f73-2b1c-4f0e-98d1-16441d128a03)
![Снимок экрана от 2023-05-25 03-01-41](https://github.com/hhru-school/segment-admin-panel-front/assets/84242545/643daf22-1cba-4d4a-93b2-0ae4eb805d6a)

*отправка*
![Снимок экрана от 2023-05-25 02-59-43](https://github.com/hhru-school/segment-admin-panel-front/assets/84242545/8f54e34a-1463-4d49-822f-102699b9de32)

**Изменения**
![Снимок экрана от 2023-05-25 03-00-30](https://github.com/hhru-school/segment-admin-panel-front/assets/84242545/01287bb8-8d83-4958-89fe-75b847fa99c5)
![Снимок экрана от 2023-05-25 03-00-34](https://github.com/hhru-school/segment-admin-panel-front/assets/84242545/fd2f6b1d-6253-4ba2-8d5c-ca78975132b8)
![Снимок экрана от 2023-05-25 03-01-04](https://github.com/hhru-school/segment-admin-panel-front/assets/84242545/46d88175-1fa7-4216-9223-2ae29718f5ff)

Beken-S committed:
Добавить модель segments

Beken-S committed:
Добавить компонент SegmentsTable

Beken-S committed:
SegmentPage.tsx переименовать в LayerSegmetPage.tsx

Beken-S committed:
SegmentsPage.tsx переименовать LayerSegmentsPage.tsx

Beken-S committed:
Добавить пункт 'Сегменты' в NavMenu

Beken-S committed:
FieldsSearchForm.tsx переименовать в SearchForm.tsx

Beken-S committed:
Исправить ScreensPage

Beken-S committed:
Исправить FieldsPage

Beken-S committed:
Исправить LayersPage

Beken-S committed:
CreateNewScreen.tsx переименовать в CreateNewScreenPage.tsx

Beken-S committed:
Добавить CreateNewSegmentPage

Beken-S committed:
Добавить SegmentsPage

Beken-S committed:
Добавить модель currentSegment

Beken-S committed:
Добавить SegmentPage

Beken-S committed:
Добавить в зависимости react-final-form

Beken-S committed:
Добавить модель roles

Beken-S committed:
Добавить хелпер sleep

Beken-S committed:
Добавить хелпер isMessage

Beken-S committed:
Исправить TableDataRow

Переименовал пропс dataConverter в dataRender и добавил возможность
пробрасывать экшен-функцию.

Beken-S committed:
Исправить размер иконки у кнопки поиска в SearchForm

Beken-S committed:
Добавить локализацию в тему

Beken-S committed:
Исправить цвет кнопки удаления у LightenChip

Beken-S committed:
Добавить FormLayout

Beken-S committed:
Добавить AddingSegmentForm

sabkaryan commented on src/components/AddingSegmentForm/ParentSegmentField.tsx:
```diff
@@ -0,0 +1,55 @@
+    };
+
+    return (
+        <Autocomplete
+            onChange={handleChange}
+            options={segments}
+            getOptionLabel={(option) => option.title}
```
давай помимо onChange передавать все пропсы из input филда. Там есть всякие onBlur, onFocus ещё , которые меняют внутренний стейт поля в форме (посещено, под фокусом и тп)

Beken-S replied:
Я правильно понимаю что туда надо будет тоже значение передавать или они как-то использую onChange?

sabkaryan replied:
нет, ничего передавать не надо. в onFocus ui компонента передать input.onFocus

Beken-S replied:
Исправил.

sabkaryan commented on src/components/AddingSegmentForm/index.tsx:
```diff
@@ -0,0 +1,166 @@
+
+const validator = (values: Values): Errors => {
+    const errors: Errors = {};
+    if (!values.title) {
+        errors.title = 'Пожалуйста, введите наименование сегмента.';
+    }
+    if (!values.roles || isEmpty(values.roles)) {
+        errors.roles = 'Пожалуйста, выберите роль для сегмента.';
+    }
```
тэги и родитель необязательные?

Beken-S replied:
Да

sabkaryan commented on src/components/AddingSegmentForm/index.tsx:
```diff
@@ -0,0 +1,166 @@
+        await sleep(2000);
+        // eslint-disable-next-line no-console
+        console.log(convertToRequestBody(values));
+        navigate(-1);
```
давай на конкретную страницу вести со списком сегментов. Пользователь мог и по прямой ссылке перейти

Beken-S replied:
Исправил.

sabkaryan commented on src/components/AddingSegmentForm/index.tsx:
```diff
@@ -0,0 +1,166 @@
+                                value={input.value}
+                                onBlur={input.onBlur}
+                                onChange={input.onChange}
+                                error={isMessage(meta.error) && meta.touched}
```
давай вынесем куда-нибудь в хелперы. везде повторяется isMessage(meta.error) && meta.touched
а тут https://github.com/hhru-school/segment-admin-panel-front/blob/a524ee19bf1d9568d4e63ae7578ae251b855c76d/src/components/AddingSegmentForm/index.tsx#L123
пропустил `&& meta.touched`

Beken-S replied:
Исправил.

sabkaryan commented on src/components/AddingSegmentForm/index.tsx:
```diff
@@ -0,0 +1,166 @@
+                            variant="contained"
+                            sx={{ width: 112 }}
+                            disableElevation
+                            disabled={rolesIsLoading || segmentsIsLoading || submitSucceeded}
```
наверно submitting, а не submitSucceeded
submitSucceeded -> уже редиректим

Beken-S replied:
Это ``LoadingButton`` он дизейблится когда идет загрузка, за это отвечает пропс ``loading``. Идея здесь была в том, что когда форма отправлена перед редиректом на доли секунды кнопка становится доступна для нажатия поэтому решил это добавить. Убрать?

sabkaryan replied:
ок, пусть будет

sabkaryan commented on src/components/AddingSegmentForm/RolesField.tsx:
```diff
@@ -0,0 +1,92 @@
+        } else {
+            onChange(newValue);
+        }
+    };
```
можно этот кусок вынести в хук. по сути в тэгах ты тоже самое делаешь
в хук можно передать функцию нормализации и маппер в fixedOptions, так как только они отличаются.
тогда в тэгах тоже можно использовать мапу (ключ - тэг)

Beken-S replied:
Вынес. Только в два разных хука, один для фиксированных значений полей, другой для извлечения данных из поля родительского сегмента.

Beken-S committed:
Исправить путь при отправке и отмене формы AddingSegmentForm

Beken-S committed:
Добавить хелпер extractFinalFormErrorState

Beken-S committed:
Исправить пропсы у полей AddingSegmentForm

Beken-S committed:
Исправить тайпгард isApiError

Убрал из обязатeльныx полей поле code.

Beken-S committed:
Добавить хук useParentFieldRoles

Beken-S committed:
Добавить хук useFixedOptions

Beken-S committed:
Исправить RolesField и TagsField

Beken-S committed:
Исправить лейауты

Отцентрировал кнопку на SecondaryLayout и исправил наименование. В
MainLayout исправил наименование настроил позиционирование обертками, а
не sx пропсами.

Beken-S committed:
Внести правки по страницам

Исправил ContentBox и внес изменения везде где он используется (страницу
LayerChangesPage и связанные с ней компоненты особо не правил так как она
поменяется)
Изменил сообщение об отсутствии данны на страницах SegmentPage и InfoPage.
На странице SegmentPage добавил ссылку на родительский слой.

Beken-S committed:
Исправить интерфейс Segment

Beken-S committed:
Сделать более явный отступ в TreeItemGroup

Beken-S committed:
Обновить моковые данные

Beken-S commented:
Пока исправлял страницу SegmentPage добавляя ссылку на родительский слой пришлось исправить дургие страницы.

Как поменялись скелетоны:

![Снимок экрана от 2023-05-27 03-08-29](https://github.com/hhru-school/segment-admin-panel-front/assets/84242545/d0c85e62-b349-489d-ac7a-7078d5db6980)
![Снимок экрана от 2023-05-27 03-11-33](https://github.com/hhru-school/segment-admin-panel-front/assets/84242545/a6578da2-609b-45c9-9dd3-70ddaaba19d9)

Как изменилось положение кнопки на второстепенном лейауте и как теперь выглядит страница сегмента:

![Снимок экрана от 2023-05-27 03-08-32](https://github.com/hhru-school/segment-admin-panel-front/assets/84242545/c428e619-5ef9-4cd5-ad40-8d9a29bb901f)
![Снимок экрана от 2023-05-27 03-08-37](https://github.com/hhru-school/segment-admin-panel-front/assets/84242545/be5e958c-394c-44f3-a580-c3773105a3c8)

Новое уведомление о отсутствии данных:
![Снимок экрана от 2023-05-27 03-10-09](https://github.com/hhru-school/segment-admin-panel-front/assets/84242545/a369ca54-6013-4078-8c98-a65b7390f58d)
![Снимок экрана от 2023-05-27 03-10-36](https://github.com/hhru-school/segment-admin-panel-front/assets/84242545/492138f7-fa3f-4142-b5f2-4297c53d3cf4)
![Снимок экрана от 2023-05-27 03-11-09](https://github.com/hhru-school/segment-admin-panel-front/assets/84242545/a461d7c1-34d2-4c96-893a-5ea7059fae02)

Новые отступы у вложенного списка:
![Снимок экрана от 2023-05-27 03-21-24](https://github.com/hhru-school/segment-admin-panel-front/assets/84242545/2be8942c-6455-44c6-8030-2228070543e8)
![Снимок экрана от 2023-05-27 03-08-55](https://github.com/hhru-school/segment-admin-panel-front/assets/84242545/18d7d3d9-1cc7-4b14-a37e-ceede9c3c963)

sabkaryan commented on src/components/AddingSegmentForm/TagsField.tsx:
```diff
@@ -8,29 +8,20 @@ import TextField from '@mui/material/TextField';
-            onChange(newValue);
-        }
-    };
+    const parentSegment = useParentFieldRoles();
```
думаю тут имеет смысл выносить только если делать более абстрактно что-то типа useFieldValue
и подписываться на value любого поля. но если потребуется потом чуть более гибко это делать всё равно придём к тому что мимикрируем стандартное апи. я бы пока оставил как было. но если уже вынес можно и так оставить, сам смотри

Beken-S replied:
Да я тоже подумал что это лишнее, но на всякий случай вынес. Исправлю.

Beken-S committed:
Удалить хук useParentFieldRoles
